pkg/k8s/resource/resource.go

```diff
@@ -464,44 +468,74 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 			}
 			entry := raw.(queueEntry)
 
-			var (
-				// eventDoneSentinel is a heap allocated object referenced by Done().
-				// If Done() is not called, a finalizer set on this object will be invoked
-				// which panics. If Done() is called, the finalizer is unset.
-				eventDoneSentinel = new(bool)
-				event             Event[T]
-			)
-			event.Done = func(err error) {
-				runtime.SetFinalizer(eventDoneSentinel, nil)
-				queue.eventDone(entry, err)
-				r.metricEventProcessed(event.Kind, err == nil)
-			}
-
-			// Add a finalizer to catch forgotten calls to Done().
-			runtime.SetFinalizer(eventDoneSentinel, doneFinalizer)
+			var event Event[T]
 
 			switch entry := entry.(type) {
 			case syncEntry:
```

Comment:
Kind of unrelated to the issue we're solving here, but should be documented at least: A consumer will receive a sync event regardless of whether the handling of the upserts was successful. That is, during the replay of the stuff in the store, if you fail to handle an event, you will receive the `Sync` before receiving the retried Upsert.